### PR TITLE
N°8008 - Webhooks: Failed to update triggering object on process response callback

### PR DESCRIPTION
**Pull request by accognet**
If the webhook is in synchronous mode
The callback method should not call BDUpdate(). The update is done at the end of the whole process.

### CODE REVIEW

**Review by Hipska — APPROVED** [review]

File: src/Core/Notification/Action/_ActionWebhook.php [hunk 1]
```diff
@@ -158,17 +159,24 @@ protected function _DoExecute($oTrigger, $aContextArgs, &$oLog)
...
+                    ],
+                ];
+                
+                if ($this->IsAsynchronous()) {
+                    $aHandlerParams['oTriggeringObject'] = [
+                        'class' => get_class($oTriggeringObject),
+                        'id' => $oTriggeringObject->GetKey(),
+                    ];
+                } else {
+                    $aHandlerParams['oObject'] = $oTriggeringObject;
```
Comment (Molkobain) [inline src/Core/Notification/Action/_ActionWebhook.php:175]:
```suggestion
                    $aHandlerParams['oTriggeringObject'] = $oTriggeringObject;
```

Oversight on my end

**Review by Molkobain — CHANGES_REQUESTED** [review]

**Review by eespie — APPROVED** [review]

**Review by odain-cbd — CHANGES_REQUESTED** [review]
please provide some tests on top of this fix before merging

**Review by Molkobain — APPROVED** [review]

File: tests/php-unit-tests/ActionWebhookTest.php [hunk 2]
```diff
@@ -196,4 +244,62 @@ public function GetContentTypeProvider()
...
+
+	public function GivenTriggerOnObjectCreate(DBObject $object): int
+	{
+		return $this->GivenObjectInDB(TriggerOnObjectCreate::class, [
+			'description'  => 'My description',
+			'target_class' => get_class($object)
+		]);
+	}
+
+	public function WhenObjectIsUpdated(DBObject $oDBObject)
```
Comment (odain-cbd) [inline tests/php-unit-tests/ActionWebhookTest.php:294]: I'd rename the function. to let know which field is updated. callback could not work and the test could pass if this function does below action
$oDBObject->Set('name', 'NameByCallBack');

File: tests/php-unit-tests/ActionWebhookTest.php [hunk 3]
```diff
@@ -196,4 +244,62 @@ public function GetContentTypeProvider()
...
+	{
+		$oDBObject->Set('first_name', 'François');
+		$oDBObject->DBUpdate();
+	}
+
+	public function WhenObjectIsCreated(DBObject $oDBObject)
```
Comment (odain-cbd) [inline tests/php-unit-tests/ActionWebhookTest.php:300]: same. i'd rename the function

File: tests/php-unit-tests/ActionWebhookTest.php [hunk 4]
```diff
@@ -196,4 +244,62 @@ public function GetContentTypeProvider()
...
+	{
+		static::$oCallBackWebhook = $oCallBack;
+		$iRemoteApplicationType = $this->GivenObjectInDB('RemoteApplicationType', [
+			'name' => 'My connection type',
+		]);
+
+		$iRemoteApplicationConnection= $this->GivenObjectInDB('RemoteApplicationConnection', [
+			'name' => 'Test localhost',
+			'remoteapplicationtype_id' => $iRemoteApplicationType,
+			'url' => 'http://localhost',
```
Comment (odain-cbd) [inline tests/php-unit-tests/ActionWebhookTest.php:263]: this test will make a remote call to this URL... maybe we should mock this call.

**Comment by Hipska** [comment]
Please be aware that the PHPDoc of ExecuteResponseHandler isn't correct anymore because of this change.

**Pull request merged**